addon/model.js

```diff
@@ -102,41 +103,50 @@ const retrieveFromCurrentState = computed('_topModel.currentState', function(key
   return this._topModel._internalModel.currentState[key];
 }).readOnly();
 
-// global buffer for initial properties to work around
-//  a)  can't write to `this` before `super`
-//  b)  core_object writes properties before calling `init`; this means that no
-//      CP or setknownProperty can rely on any initialization
-let initProperites = Object.create(null);
+const INIT_KEYS = {
+  store: 1,
+  _internalModel: 1,
+  _topModel: 1,
+  _parentModel: 1,
+  isError: 1,
+};
 
-export default class MegamorphicModel extends EmberObject {
+export default class MegamorphicModel {
+  constructor(properties) {
+    this.init(properties);
+  }
   init(properties) {
-    // Drop Ember.Object subclassing instead
-    super.init(...arguments);
+    this._init = false;
+    this.store = properties.store;
     this._store = properties.store;
     this._internalModel = properties._internalModel;
     this._cache = Object.create(null);
     this._schema = get(properties.store, '_schemaManager');
 
-    this._topModel = this._topModel || this;
-    this._parentModel = this._parentModel || null;
-    this._errors = new DS.Errors();
-    this._init = true;
-
-    this._flushInitProperties();
-  }
-
-  _flushInitProperties() {
-    let propertiesToFlush = initProperites;
-    initProperites = Object.create(null);
+    this._topModel = properties._topModel || this;
+    this._parentModel = properties._parentModel || null;
+    this._errors = DS.Errors.create();
+    this.isDestroying = false;
+    this.isDestroyed = false;
+    this.isError = properties.isError;
 
-    let keys = Object.keys(propertiesToFlush);
+    let keys = Object.keys(properties);
     if (keys.length > 0) {
       for (let i = 0; i < keys.length; ++i) {
         let key = keys[i];
-        let value = propertiesToFlush[key];
-        this.setUnknownProperty(key, value);
+        let value = properties[key];
+        if (INIT_KEYS[key] !== 1) {
+          set(this, key, value);
+        }
       }
     }
+
+    this._init = true;
```

Comment:
is this used anywhere now that `flushInitProperties` is gone?

Reply:
looks to be used when setting id during init: https://github.com/hjdivad/ember-m3/blob/66a28db3998b03af27ace254f738c05af22b70c8/addon/model.js#L363

It could be restructured, by putting `id` in `INIT_KEYS` and assigning manually to the module during construction. Might make sense to do so? Seems kinda strange to keep around the `_init` flag only for a reason that could be avoided.

Reply:
Ah yeah IIRC we can't trivially work around this in ember-m3 b/c of the way ember-data handles initial id